### DISCUSSION TITLE
Thread safety for VDO/AxEvents

### DISCUSSION THREAD

**Discussion by keesschollaart81** [AxisCommunications/discussions#149] [ACAP] [answered]
Hi,

When building an app with ACAP version 4 and using the VDO apis, do I have to care about my multi threading model? Eg: can I get buffer data from other threads than the thread that created the stream? Can two threads get buffer data simultaneously? I do also notice the AxEvents library needs a GMainLoop suggesting there is some synchronization required?

In general, my question is: are all Acap apis considered to be thread safe? 

Kind regards,
Kees

**Comment by shadysp** [comment, marked as answer]
The SDK APIs/Libraries has an internal reference counter . If the pattern: create library, create library, ..., call function, call function, ..., destroy library, destroy library, …, is followed the library will be thread safe. The recommended way is to call the create and destroy functions once in the main thread of the application.
